SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp

```diff
@@ -0,0 +1,1821 @@
+#endif
+};
+
+struct FSubViewDelta;
+
+ActorSystem::ActorSystem(const FSubView& InSubView, const FSubView& InTombstoneSubView, USpatialNetDriver* InNetDriver,
+						 FTimerManager* InTimerManager, SpatialEventTracer* InEventTracer)
+	: SubView(&InSubView)
+	, TombstoneSubView(&InTombstoneSubView)
+	, NetDriver(InNetDriver)
+	, TimerManager(InTimerManager)
+	, EventTracer(InEventTracer)
+{
+}
+
+void ActorSystem::Advance()
+{
+	for (const EntityDelta& Delta : TombstoneSubView->GetViewDelta().EntityDeltas)
+	{
+		if (Delta.Type == EntityDelta::ADD)
+		{
+			AActor* EntityActor = TryGetActor(
+				UnrealMetadata(SubView->GetView()[Delta.EntityId]
+								   .Components.FindByPredicate(ComponentIdEquality{ SpatialConstants::UNREAL_METADATA_COMPONENT_ID })
+								   ->GetUnderlying()));
+			if (EntityActor == nullptr)
+			{
+				continue;
+			}
+			UE_LOG(LogSpatialReceiver, Verbose, TEXT("The received actor with entity ID %lld was tombstoned. The actor will be deleted."),
+				   Delta.EntityId);
+			// We must first Resolve the EntityId to the Actor in order for RemoveActor to succeed.
+			NetDriver->PackageMap->ResolveEntityActor(EntityActor, Delta.EntityId);
+			RemoveActor(Delta.EntityId);
+		}
+	}
+
+	for (const EntityDelta& Delta : SubView->GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const AuthorityChange& Change : Delta.AuthorityLostTemporarily)
+			{
+				AuthorityLost(Delta.EntityId, Change.ComponentId);
+			}
+			for (const AuthorityChange& Change : Delta.AuthorityLost)
+			{
+				AuthorityLost(Delta.EntityId, Change.ComponentId);
+			}
+			for (const ComponentChange& Change : Delta.ComponentsAdded)
+			{
+				ApplyComponentAdd(Delta.EntityId, Change.ComponentId, Change.Data);
+				ComponentAdded(Delta.EntityId, Change.ComponentId, Change.Data);
+			}
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				ApplyComponentUpdate(Delta.EntityId, Change.ComponentId, Change.Update);
+				ComponentUpdated(Delta.EntityId, Change.ComponentId, Change.Update);
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				ComponentRemoved(Delta.EntityId, Change.ComponentId);
+				ApplyComponentAdd(Delta.EntityId, Change.ComponentId, Change.Data);
+				ComponentAdded(Delta.EntityId, Change.ComponentId, Change.CompleteUpdate.Data);
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRemoved)
+			{
+				ComponentRemoved(Delta.EntityId, Change.ComponentId);
+			}
+			for (const AuthorityChange& Change : Delta.AuthorityGained)
+			{
+				AuthorityGained(Delta.EntityId, Change.ComponentId);
+			}
+			for (const AuthorityChange& Change : Delta.AuthorityLostTemporarily)
+			{
+				AuthorityGained(Delta.EntityId, Change.ComponentId);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			PopulateDataStore(Delta.EntityId);
+			EntityAdded(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+			EntityRemoved(Delta.EntityId);
```

Comment:
This could destroy actors we have authority over, but have warped outside their worker's interest area.

Reply:
yep- great point. Need to revisit this with some care. A simple fix would be to add non-auth tags to Actor's authority queries- I think that is still technically correct.